### PR TITLE
System Tests: FIM - Synchronization process

### PR DESCRIPTION
**Pull request by CamiRomero**
|Related issue|
|---|
|[2389](https://github.com/wazuh/wazuh-qa/issues/2389)|

## Description
The test check that FIM currently work when we have a folder monitored and we performed some action like create the file, modify or delete files, the manager receives the changes

## Configuration options

In order to run the test, first the environment located in /test/system/provisioning/one_manager_agent must be enabled with:
sudo ansible-playbook -i inventory.yml playbook.yml --extra-vars='{"wazuh_branch": "4.2"}'
## Tests
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7865587/R1.zip) | 13/01/2021 | @CamiRomero  | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7865588/R2.zip)  | 13/01/2021 | @CamiRomero   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7865589/R3.zip)| 13/01/2021 | @CamiRomero   | 

## Tests

- [x] Proven that tests **pass** when they have to pass.
- [x] Proven that tests **fail** when they have to fail.

- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the Google Style for Python docstrings.

### CODE REVIEW

**Review by Deblintrake09 — DISMISSED** [review]

File: tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+# Copyright (C) 2015-2021, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+
```
Comment (damarisg) [inline tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py:5]: Add documentation

File: tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+def wait_for_fim_scan_end(folder_path):
+    try:
+        HostMonitor(inventory_path=inventory_path,
+                    messages_path=messages_path[2],
+                    tmp_path=tmp_path).run()
+    finally:
+        host_manager.run_command('wazuh-agent1', f'rm -rf {folder_path}')
+
```
Comment (damarisg) [inline tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py:48]: Could these methods be parameterized and added outside of the test to be used by other tests in the future or would it cause problems?

**Review by damarisg — CHANGES_REQUESTED** [review]

File: tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    # Delete folder
+    host_manager.run_command('wazuh-agent1', f'rm -rf {folder_path}')
+
+    try:
+        # Run the callback checks for the ossec.log
+        HostMonitor(inventory_path=inventory_path,
+                    messages_path=messages_path[0],
+                    tmp_path=tmp_path).run()
+    finally:
+        host_manager.run_command('wazuh-agent1', f'rm -rf {folder_path}')
```
Comment (damarisg) [inline tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py:129]: Is it could be a problem? Why  you remove 2 times the same file?

**Review by damarisg — CHANGES_REQUESTED** [review]

File: tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py (hunk 1 above)
Comment (damarisg) [reply tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py:5]: Also, Add the changelog.md

**Review by damarisg — CHANGES_REQUESTED** [review]

File: tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py (hunk 1 above)
Comment (CamiRomero) [reply tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py]: Done in commit [2c97b](https://github.com/wazuh/wazuh-qa/pull/2443/commits/2c97b0a71e51095651e5b519ede75b44571325a7) y [2413d](https://github.com/wazuh/wazuh-qa/pull/2443/commits/2413d0bf31fa4155181904016ba139024a71a835)

File: tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py (hunk 2 above)
Comment (CamiRomero) [reply tests/system/test_fim/test_fim_synchronization/test_fim_synchronization.py]: Done in commit [5972a](https://github.com/wazuh/wazuh-qa/pull/2443/commits/5972a0084b9b136aed65f0468b971bfbb477cc2e)

**Comment by CamiRomero** [comment]
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7926149/R1.zip) | 24/01/2022 | @CamiRomero  | 
| R2 | [:green_circle:] | 24/01/2022 | @CamiRomero   | 
| R3 | [:green_circle:]| 24/01/2022 | @CamiRomero   |

**Comment by CamiRomero** [comment]
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7941566/R1.zip) | 26/01/2022 | @CamiRomero  | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7941567/R2.zip) | 26/01/2022 | @CamiRomero   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7941568/R3.zip)| 26/01/2022 | @CamiRomero

**Review by Deblintrake09 — COMMENTED** [review]
Approved

## Tests
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7942316/R1.zip) | 13/01/2021 |  @Deblintrake09   | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7942314/R2.zip)  | 13/01/2021 | @Deblintrake09    | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7942317/R3.zip)| 13/01/2021 | @Deblintrake09    |

**Review by Deblintrake09 — APPROVED** [review]

File: tests/system/test_fim/__init__.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+def wait_for_fim_scan_end(HostMonitor, inventory_path, messages_path, tmp_path):
+    HostMonitor(inventory_path=inventory_path,
+                messages_path=messages_path,
+                tmp_path=tmp_path).run()
+
+
+def clean_logs(host_manager):
+    host_manager.clear_file(host='wazuh-manager', file_path=os.path.join(WAZUH_LOGS_PATH, 'ossec.log'))
+    host_manager.clear_file(host='wazuh-agent1', file_path=os.path.join(WAZUH_LOGS_PATH, 'ossec.log'))
```
Comment (damarisg) [inline tests/system/test_fim/__init__.py:24]: Could you use `clean_environment` method, please?

**Review by damarisg — CHANGES_REQUESTED** [review]

File: tests/system/test_fim/__init__.py (hunk 4 above)
Comment (CamiRomero) [reply tests/system/test_fim/__init__.py]: Done in commit [91ef4](https://github.com/wazuh/wazuh-qa/pull/2443/commits/91ef4414919fde73ccf875a29df9990852708b91)

File: deps/wazuh_testing/wazuh_testing/tools/system.py [hunk 5]
```diff
@@ -244,3 +244,13 @@ def run_shell(self, host: str, cmd: str, check: bool = False):
             stdout (str): The output of the command execution.
         """
         return self.get_host(host).ansible('shell', cmd, check=check)['stdout']
+
+
+def clean_environment(host_manager, target_files):
+    """Clears a series of files on target hosts managed by a host manager
+    Args:
+        host_manager (object): a host manager object with not None inventory_path
+        target_files (dict): a dictionary of tuples, each with the host and the path of the file to clear.
```
Comment (Deblintrake09) [inline deps/wazuh_testing/wazuh_testing/tools/system.py:253]:
```suggestion
        target_files (dict): a list of tuples, each with the host and the path of the file to clear.
```

**Review by Deblintrake09 — CHANGES_REQUESTED** [review]

File: deps/wazuh_testing/wazuh_testing/tools/system.py (hunk 5 above)
Comment (CamiRomero) [reply deps/wazuh_testing/wazuh_testing/tools/system.py]: Done in commit [0f4c9](https://github.com/wazuh/wazuh-qa/pull/2443/commits/0f4c91272c4266608f4ecae865da522ea77d24c6)

**Review by Deblintrake09 — APPROVED** [review]

**Comment by Deblintrake09** [comment]
## Tests
[R5.zip](https://github.com/wazuh/wazuh-qa/files/7961029/R5.zip)
[R6.zip](https://github.com/wazuh/wazuh-qa/files/7961030/R6.zip)
[R4.zip](https://github.com/wazuh/wazuh-qa/files/7961031/R4.zip)

|Test| Results | Date | By | Notes |
|--|--|--|--|--|
| test_files_cud.py | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7961029/R5.zip) | 28/01/2021 |  @Deblintrake09   | 
| test_files_cud.py | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7961030/R6.zip)  | 28/01/2021 | @Deblintrake09    | 
| test_files_cud.py | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7961031/R4.zip)| 28/01/2021 | @Deblintrake09    |

File: tests/system/provisioning/one_manager_agent/vars/configurations.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+manager_hostname: "wazuh-manager"
+
+agent1_id: "001"
+agent1_hostname: "wazuh-agent1"
+agent1_key: "1111111111111111111111111111111111111111111111111111111111111111"
+
+docker_network: "cluster_net"
+
+image: "dontpanicat/debian:buster"
```
Comment (snaow) [inline tests/system/provisioning/one_manager_agent/vars/configurations.yml:11]: Do not use third party docker images

File: tests/system/test_fim/test_files/test_files_cud.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+@pytest.mark.parametrize('case', ['add', 'modify', 'delete'])
+@pytest.mark.parametrize('folder_path', ['testdir1', 'testdir2', 'testdir3'])
+def test_file_cud(folder_path, case):
+    '''
+    The test will monitor a directory.
+    Finally, it will verify that the FIM event is generated
+    in agent and manager side.
+    '''
+    messages = messages_path[0]
```
Comment (snaow) [inline tests/system/test_fim/test_files/test_files_cud.py:75]: This `test_` documentation does not apply to Schema 2.0. 
Let's document the test as we did with other integration tests, or same you did with the module doc.

**Review by snaow — CHANGES_REQUESTED** [review]

File: tests/system/test_fim/test_files/test_files_cud.py (hunk 7 above)
Comment (CamiRomero) [reply tests/system/test_fim/test_files/test_files_cud.py]: Done in commit [2622c](https://github.com/wazuh/wazuh-qa/pull/2443/commits/2622cb303a637fddb1f350527327130d09de6342)

Comment (CamiRomero) [reply tests/system/test_fim/test_files/test_files_cud.py]: Done in commit [2622c](https://github.com/wazuh/wazuh-qa/pull/2443/commits/2622cb303a637fddb1f350527327130d09de6342)

File: tests/system/provisioning/one_manager_agent/vars/configurations.yml (hunk 6 above)
Comment (CamiRomero) [reply tests/system/provisioning/one_manager_agent/vars/configurations.yml:11]: When we applied the suggested changes, we note that there is a conflict with access to the remote host via ssh. We will open an Issue to work on that.

**Review by snaow — APPROVED** [review]

**Pull request merged**